### PR TITLE
Handle WEP deletion being not supported

### PR DESCRIPTION
**Pull request by robbrockbank**

### CODE REVIEW

**Comment by gunjan5** [comment]
Fixes https://github.com/projectcalico/cni-plugin/issues/334

File: k8s/k8s.go [hunk 1]
```diff
@@ -356,7 +361,12 @@ func CmdDelK8s(c *calicoclient.Client, ep api.WorkloadEndpointMetadata, args *sk
 
 	// Return the IPAM error if there was one. The IPAM error will be lost if there was also an error in cleaning up
 	// the device or endpoint, but crucially, the user will know the overall operation failed.
-	return ipamErr
+	if ipamErr != nil {
```
Comment (caseydavenport) [inline k8s/k8s.go]: Why is this change required?

File: k8s/k8s.go [hunk 2]
```diff
@@ -340,6 +340,11 @@ func CmdDelK8s(c *calicoclient.Client, ep api.WorkloadEndpointMetadata, args *sk
 			// Returning an error here is with the assumption that k8s (kubelet) retries deleting again.
 			logger.WithField("endpoint", wep).Warning("Error deleting endpoint: endpoint was modified before it could be deleted.")
 			return fmt.Errorf("Error deleting endpoint: endpoint was modified before it could be deleted: %v", err)
+		case cerrors.ErrorOperationNotSupported:
+			// If the operation is not supported, then we must be running with KDD.  Ignore this error
+			// and continue - the WEP is actually backed by the Pod and the deletion will be handled
```
Comment (gunjan5) [inline k8s/k8s.go]: actually -> ~acctually~

File: k8s/k8s.go (hunk 1 above)
Comment (robbrockbank) [inline k8s/k8s.go]: it was there to get the extra log in

File: k8s/k8s.go [hunk 3]
```diff
@@ -340,6 +340,11 @@ func CmdDelK8s(c *calicoclient.Client, ep api.WorkloadEndpointMetadata, args *sk
 			// Returning an error here is with the assumption that k8s (kubelet) retries deleting again.
 			logger.WithField("endpoint", wep).Warning("Error deleting endpoint: endpoint was modified before it could be deleted.")
 			return fmt.Errorf("Error deleting endpoint: endpoint was modified before it could be deleted: %v", err)
+		case cerrors.ErrorOperationNotSupported:
+			// If the operation is not supported, then we must be running with KDD.  Ignore this error
```
Comment (gunjan5) [inline k8s/k8s.go]: how about `Delete operation is not supported for KDD backend.`

Comment (gunjan5) [inline k8s/k8s.go]: also don't forget the 2 spaces after `.` :)

File: k8s/k8s.go [hunk 4]
```diff
@@ -356,7 +361,12 @@ func CmdDelK8s(c *calicoclient.Client, ep api.WorkloadEndpointMetadata, args *sk
 
 	// Return the IPAM error if there was one. The IPAM error will be lost if there was also an error in cleaning up
 	// the device or endpoint, but crucially, the user will know the overall operation failed.
-	return ipamErr
+	if ipamErr != nil {
+		return ipamErr
+	}
+
+	logger.WithField("endpoint", wep).Info("Endpoint deletion complete.")
```
Comment (caseydavenport) [inline k8s/k8s.go]: I don't think this is the right log - Endpoint deletion was complete way up above, this log should say that all of tearDown is complete.

File: k8s/k8s.go [hunk 5]
```diff
@@ -340,6 +340,10 @@ func CmdDelK8s(c *calicoclient.Client, ep api.WorkloadEndpointMetadata, args *sk
 			// Returning an error here is with the assumption that k8s (kubelet) retries deleting again.
 			logger.WithField("endpoint", wep).Warning("Error deleting endpoint: endpoint was modified before it could be deleted.")
 			return fmt.Errorf("Error deleting endpoint: endpoint was modified before it could be deleted: %v", err)
+		case cerrors.ErrorOperationNotSupported:
+			// KDD does not support WorkloadEndpoint deletion, the WEP is backed by the Pod and the
+                        // deletion will be handled by Kubernetes. This error can be ignored.
```
Comment (gunjan5) [inline k8s/k8s.go]: weird indent

**Comment by gunjan5** [comment]
ok manual test looks good. JGOC!
